### PR TITLE
[JAX] Context Parallel Attention with All-Gather

### PR DESCRIPTION
**Pull request by mgoldfarb-nvidia**
# Description

Adds support for context parallel fused attention based on an all-gather/reduce-scatter approach. This implementation exposes the collective communication between CP ranks.

The first implementation of CP only support causal and no masking without bias. Additional QKV formats and configurations will be added to subsequent PRs.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Adds context parallel axis resource and new primitives to fused attention.

# Checklist:

- [X] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by ptrendx** [comment]
How is this different from #1059?

**Comment by mgoldfarb-nvidia** [comment]
The other method is implemented as ring attention with point to point comms. Both will ultimately provide context parallel attention but some forms of attention e.g window attention are more easily supported with ag approach. We also may get better scaling of the ag comms on multi node setups.

Michael Goldfarb
Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by ptrendx** [comment]
Hmmm, sure, but it still feels like duplicate of the functionality. Could you maybe collaborate with @mingxu1067 to merge your work with his?

**Comment by mgoldfarb-nvidia** [comment]
Sure thing. Ming and I have already been in discussion on how to merge the PRs. Likely there should be an initial CP attention PR and will need to follow with updated to other components on the jax side as we implement the rest of the CP features for jax.

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 1]
```diff
@@ -1020,6 +895,361 @@ def sharded_impl(
...
+
+            return dq, dk, dv, dbias
+
+        return mesh, impl, out_shardings, arg_shardings
+
+
+register_primitive(FusedAttnCPWithAllGatherBwdPrimitive)
+
+
 def fused_attn_fwd(
```
Comment (mgoldfarb-nvidia) [inline transformer_engine/jax/cpp_extensions/attention.py:1253]: This is an important discussion point: Do we keep CP hidden behind a common function or add a separate flavor of `fused_attn_cp_allgather_fwd`. The though here was it makes sense to keep a common interface that naturally support CP. Other implementation strategies e.g. ring can be exposed via argument.

File: transformer_engine/jax/attention.py [hunk 2]
```diff
@@ -155,6 +167,75 @@ def _obtain_batch_and_max_seqlen(qkv, qkv_layout):
...
+        )
+    )
+
+    parts = []
+    if not inverse:
+        for cp_rank in range(cp_size):
+            # [B, S, H, D]: [B, 2*cp_size, S/2*cp_size, H, D] -> [B, 2, S/2*cp_size, H, D]
+            # [S, B, H, D]: [2*cp_size, S/2*cp_size, B, H, D] -> [2, S/2*cp_size, B, H, D]
+            index = jnp.array([cp_rank, (2 * cp_size - cp_rank - 1)])
+            parts.append(jnp.take(tensor, index, axis=seq_dim))
```
Comment (zlsh80826) [inline transformer_engine/jax/attention.py:207]: Is it possible to use `take_along_axis`? Not sure if it can produce better results.

File: transformer_engine/jax/sharding.py [hunk 3]
```diff
@@ -148,6 +152,22 @@ def num_of_devices():
...
+    If the mesh is None, it would be replaced
+    by the global mesh.
+    """
+    if mesh is None:
+        mesh = _PXLA_THREAD_RESOURCES.env.physical_mesh
+
+    if axis is None:
+        return 1
+
+    assert axis in mesh.shape, f"{axis} is not a axis of the given mesh f{mesh.shape}"
```
[the remark this answers is not in the document]
Comment (mgoldfarb-nvidia) [reply transformer_engine/jax/sharding.py]: Yes.

File: transformer_engine/jax/attention.py [hunk 4]
```diff
@@ -155,6 +167,75 @@ def _obtain_batch_and_max_seqlen(qkv, qkv_layout):
...
+        case QKVFormat.BSHD:
+            seq_dim = 1
+        case _:
+            raise ValueError(f"{tensor_format=} is not supported for causal load balancing.")
+
+    if cp_size == 1:
+        return tensor
+
+    if cp_size % 2 != 0:
+        raise ValueError(f"{cp_size=} must be a power of 2.")
```
[the remark this answers is not in the document]
Comment (mgoldfarb-nvidia) [reply transformer_engine/jax/attention.py]: Good catch - multiple of 2.

File: transformer_engine/jax/attention.py (hunk 2 above)
Comment (mgoldfarb-nvidia) [reply transformer_engine/jax/attention.py:207]: Switched to `take_along_axis`

File: tests/jax/test_distributed_fused_attn.py [hunk 5]
```diff
@@ -310,3 +320,193 @@ def ref_func(query, kv, mask):
...
+                        print("std: ", jnp.std(target_grads[i]), jnp.std(ref_grads[i]))
+                        print("var: ", jnp.var(target_grads[i]), jnp.var(ref_grads[i]))
+                        print("max diff: ", jnp.max(jnp.abs(target_grads[i] - ref_grads[i])))
+                        raise e
+
+
+class TestReorderCausalLoadBalancing:
+    @pytest.mark.parametrize("cp_size", [2, 4, 8])
+    @pytest.mark.parametrize("shape", [[1, 16, 1, 1], [4, 32, 12, 32], [3, 32, 8, 64]])
+    @pytest.mark.parametrize("dtype", DTYPES)
```
Comment (zlsh80826) [inline tests/jax/test_distributed_fused_attn.py:497]: We can just testing `bfloat16`. There are many configurations for the flash attention and may cause the time out on CI. The test_fused_attn.py also only tested the bfloat16 data type.

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 6]
```diff
@@ -1020,6 +897,378 @@ def sharded_impl(
...
+                results = []
+                for sub_idx in range(2):
+                    if config.attn_mask_type == NVTE_Mask_Type.NVTE_NO_MASK:
+                        k_unmasked, v_unmasked = k, v  # full kv used for unmasked
+                    else:
+                        k_unmasked, v_unmasked = helper.slice_kv(k, v, kv_seqlens_for_rank[sub_idx])
+
+                    q_seqlen = q_seqlen / (cp_size * 2)
+                    num_kv_chunks = kv_max_seqlen // kv_seqlens_for_rank[sub_idx]
+                    kv_seqlen = (kv_seqlen / (cp_size * 2)) * num_kv_chunks
```
Comment (zlsh80826) [inline transformer_engine/jax/cpp_extensions/attention.py:1090]: the kv_seqlen will be override in the first loop and reused in the second loop, is this expected?

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 7]
```diff
@@ -1020,6 +897,378 @@ def sharded_impl(
...
+            # to select the appropriate computation. Each case generates a [..., SEQ/CP, ..] tensor
+            # meeting the expectation of the SPMD model.
+            # TODO(mgoldfarb-nvidia): When cuDNN supports we should be able to make use of a padding
+            # mask/sequence length tensor to avoid this unrolled loop.
+            def _cross_attn(idx, q, k, v, bias, q_seqlen, kv_seqlen, seed):
+                kv_max_seqlen = k.shape[1]
+                kv_seqlen_per_subrank = kv_max_seqlen // (cp_size * 2)
+                assert kv_max_seqlen % cp_size == 0, "sequence length must evenly divide cp size"
+
+                q_split = jnp.split(q, 2, axis=1)
```
Comment (zlsh80826) [inline transformer_engine/jax/cpp_extensions/attention.py:1075]: I think the reason for spliting q here is for load balancing, right? When load balancing is enabled, is it the user's responsibility to reorder the q sequence?

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 8]
```diff
@@ -1020,6 +896,378 @@ def sharded_impl(
...
+        if self.config.attn_mask_type == NVTE_Mask_Type.NVTE_CAUSAL_MASK:
+            return NVTE_Mask_Type.NVTE_CAUSAL_BOTTOM_RIGHT_MASK
+        return self.config.attn_mask_type
+
+    def all_gather_kv(self, k, v):
+        """Performs a all-gather of k and v over context parallel ranks."""
+
+        def ag(x):
+            return lax_paral_op(
+                x, lax.all_gather, self.config.cp_axis, mesh=self.mesh, axis=1, tiled=True
```
[the remark this answers is not in the document]
Comment (mgoldfarb-nvidia) [reply transformer_engine/jax/cpp_extensions/attention.py:944]: Currently we've only implemented BSHD layouts. I was planning to separate the other layout support i.e. BHSD and THD in separate PRs. Should we plan to add them here?

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 9]
```diff
@@ -1020,6 +896,378 @@ def sharded_impl(
...
+    Fused Attention Backward with Context Parallelism Primitive.
+
+    This context parallel implementation uses all-gather to collect KV and dKV inputs from context parallel ranks.
+    The gradients are subsequently reduce-scattered back to each context parallel rank.
+    """
+
+    @staticmethod
+    def partition(config, mesh, arg_infos, result_infos):
+        # Call base implementation for non-context parallel mesh to avoid unecessary work.
+        is_context_parallel = get_mesh_axis_size(config.cp_axis, mesh)
```
[the remark this answers is not in the document]
Comment (mgoldfarb-nvidia) [reply transformer_engine/jax/cpp_extensions/attention.py]: This was a bug and got fixed.

File: transformer_engine/jax/cpp_extensions/attention.py (hunk 7 above)
Comment (mgoldfarb-nvidia) [reply transformer_engine/jax/cpp_extensions/attention.py:1075]: Yes this is part of load balance the cp ranks. We split the entire set of sequences into `CP*2` groups. The user needs to call `reorder_causal_load_balancing` prior to attention or ideally at the beginning of a transformer layer stack.

Screenshot 2024-08-15 at 10 38 39 AM

File: tests/jax/test_distributed_fused_attn.py [hunk 10]
```diff
@@ -310,3 +320,193 @@ def ref_func(query, kv, mask):
...
+                    mask=mask,
+                    seed=None,
+                    attn_bias_type=attn_bias_type,
+                    attn_mask_type=attn_mask_type,
+                    qkv_layout=qkv_layout,
+                    scaling_factor=scaling_factor,
+                    dropout_probability=dropout_prob,
+                    is_training=is_training,
+                    context_parallel_causal_load_balanced=load_balanced,
+                ),
```
[the remark this answers is not in the document]
Comment (mgoldfarb-nvidia) [reply tests/jax/test_distributed_fused_attn.py:424]: Done.

File: tests/jax/test_distributed_fused_attn.py (hunk 5 above)
Comment (mgoldfarb-nvidia) [reply tests/jax/test_distributed_fused_attn.py]: Done.

File: transformer_engine/jax/cpp_extensions/attention.py (hunk 6 above)
Comment (mgoldfarb-nvidia) [reply transformer_engine/jax/cpp_extensions/attention.py]: Good catch. This has been fixed.

File: transformer_engine/jax/cpp_extensions/attention.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (zlsh80826) [reply transformer_engine/jax/cpp_extensions/attention.py:944]: It is good for now, I think we have enough protect for not accepting `SBHD`. Thanks.

**Review by zlsh80826 — APPROVED** [review]
LGTM! Thanks for the contributions, especially the `_FusedAttnConfig`, that's quite simplify the argument passing.

**Comment by zlsh80826** [comment]
Please help check the pre-commit failure. You can run `pre-commit install` in your working directory, then it will check when each time you commit.
